### ISSUE TITLE
Howto install Qira on ubuntu 20.04

### ISSUE THREAD

**Issue by smurfd** [geohot/qira#258]
This is done on a fresh install of ubuntu 20.04 with installed git.
grab these :
[qira-qemu.txt](https://github.com/geohot/qira/files/6331368/qira-qemu.txt)
[qira.txt](https://github.com/geohot/qira/files/6331369/qira.txt)

git clone https://github.com/geohot/qira.git
cd qira
patch -p1 < ../qira.txt

install.sh will end in error : util/memfd.c:43:12: error: static declaration of ‘memfd_create’ follows non-static declaration

cd tracers/qemu/qemu
patch -p1 < ../../../../qira-qemu.txt

cd ../../..
./install.sh
Booom :)

**Comment by Wgmlgz** [comment]
I tried  to use this but I get this error:

    ERROR: Command errored out with exit status 1:
     command: /usr/bin/python2 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-D2fxo9/filelock/setup.py'"'"'; __file__='"'"'/tmp/pip-install-D2fxo9/filelock/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-pip-egg-info-NbW9l7                                                                                         
         cwd: /tmp/pip-install-D2fxo9/filelock/                                                                                                                      
    Complete output (8 lines):                                                                                                                                       
    /usr/lib/python2.7/distutils/dist.py:267: UserWarning: Unknown distribution option: 'test_suite'                                                                 
      warnings.warn(msg)                                                                                                                                             
    usage: setup.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]                                                                                            
       or: setup.py --help [cmd1 cmd2 ...]                                                                                                                           
       or: setup.py --help-commands                                                                                                                                  
       or: setup.py cmd --help                                                                                                                                       
                                                                                                                                                                     
    error: invalid command 'egg_info'                                                                                                                                
    ----------------------------------------                                                                                                                         
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.

Can you please help me, I am n00b(

**Comment by smurfd** [comment]
weird, try to in the qira folder:
mv venv venv1
then rerun the install.sh

if that does not work, im not sure ...maby you had installed pip2 already!?

**Comment by Wgmlgz** [comment]
Yes, a have pip2 installed (I comment pip2 installation lines), but now I am getting another error:

util/memfd.c:43:12: error: static declaration of ‘memfd_create’ follows non-static declaration
   43 | static int memfd_create(const char *name, unsigned int flags)
      |            ^~~~~~~~~~~~
In file included from /usr/include/x86_64-linux-gnu/bits/mman-linux.h:113,
                 from /usr/include/x86_64-linux-gnu/bits/mman.h:34,
                 from /usr/include/x86_64-linux-gnu/sys/mman.h:41,
                 from /home/wgmlgz/qira/tracers/qemu/qemu/include/qemu/osdep.h:142,
                 from util/memfd.c:28:
/usr/include/x86_64-linux-gnu/bits/mman-shared.h:50:5: note: previous declaration of ‘memfd_create’ was here
   50 | int memfd_create (const char *__name, unsigned int __flags) __THROW;
      |     ^~~~~~~~~~~~
make: *** [/home/wgmlgz/qira/tracers/qemu/qemu/rules.mak:57: util/memfd.o] Error 1
make: *** Waiting for unfinished jobs....

**Comment by smurfd** [comment]
from my 1st post : 
```
install.sh will end in error : util/memfd.c:43:12: error: static declaration of ‘memfd_create’ follows non-static declaration

cd tracers/qemu/qemu
patch -p1 < ../../../../qira-qemu.txt

cd ../../..
./install.sh
```

**Comment by Wgmlgz** [comment]
I ended up just reinstalling my ubuntu vm and your patches are works! By the way, can you please make patch for kali?

**Comment by smurfd** [comment]
same works for kali linux also, u just need to grab these and install them before ... since they seem to have screwed up something about their python2 install.
http://ftp.us.debian.org/debian/pool/main/p/python-scandir/python-scandir_1.9.0-2_amd64.deb
http://ftp.us.debian.org/debian/pool/main/p/python-filelock/python-filelock_3.0.10-1_all.deb

**Comment by Wgmlgz** [comment]
Man, that's works for kali! Thanks you a lot!

**Comment by smurfd** [comment]
These will not force python2, but use python3 only

Grab and save to a folder. had to update some stuff in the requirements.txt

[qira-py39.txt](https://github.com/geohot/qira/files/6372824/qira-py39.txt)
[qira-qemu-py39.txt](https://github.com/geohot/qira/files/6372825/qira-qemu-py39.txt)

Then : 
```
git clone https://github.com/geohot/qira
cd qira/tracers/qemu
git clone https://github.com/geohot/qemu.git --depth 1 --branch qira
cd qemu
patch -p1 < ../../../../qira-qemu-py39.patch
cd ../../..
patch -p1 < ../qira-py39.patch
./install.sh
```
OBVIOUSLY there might be some subtleties that im missing or have broken. use at own risk, same ofcourse goes for previously pasted patches...


Sidenote:
Also tried to build against qemu git main. the trick to building that was to remove the `--cpu=unknown` and rearrange the symlinks in tracers/qemu to point to point like : qira-aarch64 -> qemu/build/aarch64-linux-user/qemu-aarch64

Something with the webserver/client/update part did not like this however :/
pretty sure it is some python2 to python3 junk that breakts things.

**Comment by ManasNagaraj** [comment]
Apparently python-pip has been removed from the repositories for Ubuntu 20.04 and according to the install.sh script python-pip is being used may be that's the reason
[Ubuntu 20.04](https://i.stack.imgur.com/lagdO.png)

**Comment by williams4564** [comment]
Thank you @smurfd your patch works on my Ubuntu 20.04 VM

**Comment by a1eaiactaest** [comment]
works on Ubuntu 21.04 as well.

**Comment by a1eaiactaest** [comment]
I created a fork with all patches applied. if someone doesn't want to manually patch qira and qemu just clone [a1eaiactaest/qira](https://github.com/a1eaiactaest/qira.git) and normally run `./install.sh`. i think everything should be fine.

**Comment by IdanBanani** [comment]
> I created a fork with all patches applied. if someone doesn't want to manually patch qira and qemu just clone [a1eaiactaest/qira](https://github.com/a1eaiactaest/qira.git) and normally run `./install.sh`. i think everything should be fine.

WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
./install.sh: line 15: tracers/qemu/qira-i386: No such file or directory
building QEMU
./qemu_build.sh: line 10: ./configure: No such file or directory

**Comment by IdanBanani** [comment]
@smurfd Do you have any idea if the integration (synchronization) with IDA requires an IDA PRO version (to use the plugin) to be installed on Linux or it is possible to use an IDA PRO on Windows which uses remote debugging connection?
Nevertheless, even without IDA this could be useful (too bad the project is kind of dead).

**Comment by smurfd** [comment]
> @smurfd Do you have any idea if the integration (synchronization) with IDA requires an IDA PRO version (to use the plugin) to be installed on Linux or it is possible to use an IDA PRO on Windows which uses remote debugging connection? Nevertheless, even without IDA this could be useful (too bad the project is kind of dead).

sorry i do not have a clue unfortunately. :(

**Comment by BelaBartok39** [comment]
> > I created a fork with all patches applied. if someone doesn't want to manually patch qira and qemu just clone [a1eaiactaest/qira](https://github.com/a1eaiactaest/qira.git) and normally run `./install.sh`. i think everything should be fine.
> 
> WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv ./install.sh: line 15: tracers/qemu/qira-i386: No such file or directory building QEMU ./qemu_build.sh: line 10: ./configure: No such file or directory

I'm having the same issue. Looks for ./configure file that does not exist.
